keycloak-event-emitter/src/main/java/io/cloudtrust/keycloak/kafkaeventemitter/KafkaEventEmitterProvider.java

```diff
@@ -0,0 +1,172 @@
+    }
+
+    @Override
+    public void onEvent(AdminEvent adminEvent, boolean includeRepresentation) {
+        long uid = idGenerator.nextValidId();
+        IdentifiedAdminEvent identifiedAdminEvent = new IdentifiedAdminEvent(uid, adminEvent);
+        ExtendedAdminEvent customAdminEvent = completeAdminEventAttributes(identifiedAdminEvent);
+
+        // Flatbuffer serialization
+        ByteBuffer buffer = SerialisationUtils.toFlat(customAdminEvent);
+
+        produceEvent(buffer, customAdminEvent.getAuthDetails().getUserId(),adminEventTopic);
+    }
+
+    @Override
+    public void close() {
+
+    }
+
+    private void completeEventAttributes(Event event) {
+        // add username if missing
+        if (event.getDetails() == null) {
+            event.setDetails(new HashMap<>());
+        }
+        String eventUsername = event.getDetails().get(Details.USERNAME);
+        if (!Strings.isNullOrEmpty(event.getUserId()) && Strings.isNullOrEmpty(eventUsername)) {
+            findUser(event.getUserId(), event.getRealmId(), u -> event.getDetails().put(Details.USERNAME, u.getUsername()));
+        }
+    }
+
+    private ExtendedAdminEvent completeAdminEventAttributes(IdentifiedAdminEvent adminEvent) {
+        ExtendedAdminEvent extendedAdminEvent = new ExtendedAdminEvent(adminEvent);
+        // add always missing agent username
+        ExtendedAuthDetails extendedAuthDetails = extendedAdminEvent.getAuthDetails();
+        if (!Strings.isNullOrEmpty(extendedAuthDetails.getUserId())) {
+            findUser(extendedAuthDetails.getUserId(), extendedAuthDetails.getRealmId(), u -> extendedAuthDetails.setUsername(u.getUsername()));
+        }
+        // add username if resource is a user
+        String resourcePath = extendedAdminEvent.getResourcePath();
+        if (resourcePath != null && resourcePath.startsWith("users")) {
+            // parse userID
+            String pattern = "^users/([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})$";
+            Pattern r = Pattern.compile(pattern);
+            Matcher m = r.matcher(resourcePath);
+            if (m.matches()) {
+                String userId = m.group(1);
+                extendedAdminEvent.getDetails().put("target_user_id", userId);
+                findUser(userId, adminEvent.getRealmId(), u -> extendedAdminEvent.getDetails().put("target_username", u.getUsername()));
+            }
+        }
+
+        return extendedAdminEvent;
+    }
+
+    private void findUser(String userId, String realmId, Consumer<UserModel> whenUserFound) {
+        RealmModel realm = keycloakSession.realms().getRealm(realmId);
+        if (realm != null) {
+            UserModel user = keycloakSession.users().getUserById(realm, userId);
+
+            if (user != null) {
+                whenUserFound.accept(user);
+            }
+        }
+    }
+
+    private void produceEvent(ByteBuffer buffer, String key,String topic) {
```

Comment:
if state == STARTED 
producer.send // cas normal
return

if producer == null
lock{
add event in queue
state= PENDING_EVENTS
}
retrun

if state == PENDING_EVENTS
lock{
consumme all events
state = STARTED
}
return

All cases should have been considered
Should not happen, throw exception